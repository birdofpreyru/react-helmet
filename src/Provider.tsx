import {
  type FunctionComponent,
  type ReactNode,
  createContext,
  useRef,
} from 'react';

import type {
  ContextValue,
  HelmetDataContext,
  HelmetProps,
  HelmetProviderHeap,
} from './types';

import { newServerState } from './server';
import { IS_DOM_ENVIRONMENT } from './constants';
import { calcAggregatedState } from './utils';
import { commitTagChanges } from './client';

export const Context = createContext<ContextValue | undefined>(undefined);

type ProviderProps = {
  children?: ReactNode;
  context?: HelmetDataContext;
};

const HelmetProvider: FunctionComponent<ProviderProps> = ({
  children,
  context,
}) => {
  const { current: heap } = useRef<HelmetProviderHeap>({
    firstRender: true,
    helmets: [],
    state: undefined,
  });

  const contextValueRef = useRef<ContextValue>(null);

  contextValueRef.current ??= {
    clientApply() {
      if (IS_DOM_ENVIRONMENT && !heap.state) {
        heap.state = calcAggregatedState(heap.helmets);
        if (heap.state.defer) {
          heap.nextAnimFrameId ??= requestAnimationFrame(() => {
            heap.state ??= calcAggregatedState(heap.helmets);
            commitTagChanges(heap.state, heap.firstRender);
            heap.firstRender = false;
            delete heap.nextAnimFrameId;
          });
        } else {
          if (heap.nextAnimFrameId !== undefined) {
            cancelAnimationFrame(heap.nextAnimFrameId);
            delete heap.nextAnimFrameId;
          }
          commitTagChanges(heap.state, heap.firstRender);
          heap.firstRender = false;
        }
      }
    },
    update(id: string, props: HelmetProps | undefined) {
      const idx = heap.helmets.findIndex((item) => item[0] === id);
      if (idx >= 0) {
        delete heap.state;
        if (props) heap.helmets[idx]![1] = props;
        else heap.helmets.splice(idx, 1);
      } else if (props) {
        delete heap.state;
        heap.helmets.push([id, props]);
      }
    },
  };

  if (context && (!context.helmet || context.helmet !== heap.serverState)) {
    heap.serverState ??= newServerState(heap);

    // eslint-disable-next-line no-param-reassign
    context.helmet = heap.serverState;
  }

  return <Context value={contextValueRef.current}>{children}</Context>;
};

export default HelmetProvider;
