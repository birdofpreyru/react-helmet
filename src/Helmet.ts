import {
  type FunctionComponent,
  type ReactElement,
  type ReactNode,
  Children,
  use,
  useEffect,
  useId,
} from 'react';

import { REACT_TAG_MAP, TAG_NAMES, VALID_TAG_NAMES } from './constants';

import { Context } from './Provider';

import type {
  BaseProps,
  BodyProps,
  HelmetChildProps,
  HelmetProps,
  HtmlProps,
  LinkProps,
  MetaProps,
  NoscriptProps,
  ScriptProps,
  StyleProps,
  TitleProps,
} from './types';
import { cloneProps, mergeProps, pushToPropArray } from './utils';

function assertChildType(
  childType: ReactElement['type'],
  nestedChildren: ReactNode,
): asserts childType is TAG_NAMES {
  if (typeof childType !== 'string') {
    throw Error(
      'You may be attempting to nest <Helmet> components within each other, which is not allowed. Refer to our API for more information.',
    );
  }

  if (!(VALID_TAG_NAMES as string[]).includes(childType)) {
    throw Error(
      `Only elements types ${VALID_TAG_NAMES.join(', ')} are allowed. Helmet does not support rendering <${childType}> elements. Refer to our API for more information.`,
    );
  }

  if (
    !nestedChildren
    || typeof nestedChildren === 'string'
    || Array.isArray(nestedChildren)
    // TODO: This piece of the check is wrong when parent is a fragment,
    // and thus children may not be an array of strings.
    // && nestedChildren.every((item) => typeof item === 'string')
  ) return;

  throw Error(
    `Helmet expects a string as a child of <${childType}>. Did you forget to wrap your children in braces? ( <${childType}>{\`\`}</${childType}> ) Refer to our API for more information.`,
  );
}

/**
 * Given a string key, it checks it against the legacy mapping between supported
 * HTML attribute names and their corresponding React prop names (for the names
 * that are different). If found in the mapping, it prints a warning to console
 * and returns the mapped prop name. Otherwise, it just returns the key as is,
 * assuming it is already a valid React prop name.
 */
function getPropName(key: string): keyof HelmetChildProps {
  const res = REACT_TAG_MAP[key];
  if (res) {
    // eslint-disable-next-line no-console
    console.warn(
      `"${key}" is not a valid JSX prop, replace it by "${res}"`,
    );
  }
  return (res ?? key) as keyof HelmetChildProps;
}

/**
 * Given children and props of a <Helmet> component, it reduces them to a single
 * props object.
 *
 * TODO: I guess, it should be further refactored, to make it cleaner...
 * though, it should perfectly work as is, so not a huge priority for now.
 */
function reduceChildrenAndProps(props: HelmetProps): Omit<HelmetProps, 'children'> {
  // NOTE: `props` are clonned, thus it is safe to push additional items to
  // array values of `res`, and to re-assign non-array values of `res`, without
  // the risk to mutate the original `props` object.
  const res: HelmetProps = cloneProps(props);

  // TODO: This is a temporary block, for compatibility with legacy library.
  for (const item of Object.values(props)) {
    if (Array.isArray(item)) {
      for (const it of item) {
        // TODO: This condition is actually needed to prevent some test failures,
        // I guess, something is messed up with related types?
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
        if (it) {
          for (const key of Object.keys(it)) {
            const p = getPropName(key);
            if (p !== key) {
              it[p] = it[key as keyof HelmetChildProps] as unknown;
              delete it[key as keyof HelmetChildProps];
            }
          }
        }
      }
    } else if (item && typeof item === 'object') {
      const it = item as HelmetChildProps;
      for (const key of Object.keys(it)) {
        const p = getPropName(key);
        if (p !== key) {
          it[p] = it[key as keyof HelmetChildProps] as unknown;
          delete it[key as keyof HelmetChildProps];
        }
      }
    }
  }

  // eslint-disable-next-line complexity
  Children.forEach(props.children, (child) => {
    if (child === undefined || child === null) return;

    if (typeof child !== 'object' || !('props' in child)) {
      throw Error(
        `"${typeof child}" is not a valid <Helmet> descendant`,
      );
    }

    let nestedChildren: ReactNode;
    const childProps: HelmetChildProps = {};
    if (child.props) {
      for (const [key, value] of Object.entries(child.props)) {
        if (key === 'children') nestedChildren = value as ReactNode;
        else childProps[getPropName(key)] = value as unknown;
      }
    }

    let { type } = child;
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-conversion
    if (typeof type === 'symbol') type = (type as 'symbol').toString();
    assertChildType(type, nestedChildren);

    function assertStringChild(child2: ReactNode): asserts child2 is string {
      if (typeof child2 !== 'string') {
        // TODO: We want to throw, but the legacy code did not, so we won't for
        // now.
        // eslint-disable-next-line no-console
        console.error(`child of ${type as string} element should be a string`);

        /*
        throw Error(
          // NOTE: assertChildType() above guarantees that `type` is a string,
          // although it is not expressed in a way TypeScript can automatically
          // pick up.
        );
        */
      }
    }

    switch (type) {
      case TAG_NAMES.BASE:
        res.base = childProps as BaseProps;
        break;

      case TAG_NAMES.BODY:
        res.bodyAttributes = childProps as BodyProps;
        break;

      case TAG_NAMES.FRAGMENT:
        mergeProps(res, reduceChildrenAndProps({ children: nestedChildren }));
        break;

      case TAG_NAMES.HTML:
        res.htmlAttributes = childProps as HtmlProps;
        break;

      case TAG_NAMES.LINK:
      case TAG_NAMES.META:
        if (nestedChildren) {
          throw Error(
            `<${type} /> elements are self-closing and can not contain children. Refer to our API for more information.`,
          );
        }
        pushToPropArray(res, type, childProps as LinkProps | MetaProps);
        break;

      case TAG_NAMES.NOSCRIPT:
      case TAG_NAMES.SCRIPT:
        if (nestedChildren !== undefined) {
          assertStringChild(nestedChildren);
          (childProps as NoscriptProps | ScriptProps)
            .innerHTML = nestedChildren;
        }
        pushToPropArray(res, type, childProps);
        break;

      case TAG_NAMES.STYLE:
        assertStringChild(nestedChildren);
        (childProps as StyleProps).cssText = nestedChildren;
        pushToPropArray(res, type, childProps as StyleProps);
        break;

      case TAG_NAMES.TITLE:
        res.titleAttributes = childProps as TitleProps;

        if (typeof nestedChildren === 'string') res.title = nestedChildren;

        // When title contains {} expressions the children are an array of
        // strings, and other values.
        else if (Array.isArray(nestedChildren)) res.title = nestedChildren.join('');
        break;

      case TAG_NAMES.HEAD:
      default: {
        // TODO: Perhaps, we should remove HEAD entry from TAG_NAMES?
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const bad: TAG_NAMES.HEAD = type;
      }
    }
  });

  delete res.children;
  return res;
}

const Helmet: FunctionComponent<HelmetProps> = (props) => {
  const context = use(Context);

  if (!context) {
    throw Error(
      '<Helmet> component must be within a <HelmetProvider> children tree',
    );
  }

  const id = useId();

  // TODO: Agh... we need it here to ensure that it works server-side,
  // and we need the same in the useEffect() below to ensure it works
  // client-side in strict mode (and, thus completely correctly from React's
  // pure component / side effect logic). It clearly should be optimized,
  // but let's care about it later.
  context.update(id, reduceChildrenAndProps(props));

  // TODO: I guess, these two useEffect() can be merged together, which should
  // also allow to simplify and optimize the client-side management of attrs and
  // elements managed by these. Though, keeping them separate is an easier way
  // for now to ensure backward compatibility.
  useEffect(() => {
    context.update(id, reduceChildrenAndProps(props));
    context.clientApply();
  });

  useEffect(() => () => {
    context.update(id, undefined);
  }, [context, id]);

  return null;
};

export default Helmet;
