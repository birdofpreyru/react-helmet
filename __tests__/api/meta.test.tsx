/** @jest-environment jsdom */

import { Helmet } from '../../src';
import { HELMET_ATTRIBUTE } from '../../src/constants';
import { renderClient } from '../../jest/browser-utils';

describe('meta tags', () => {
  describe('API', () => {
    it('updates meta tags', () => {
      renderClient(
        <Helmet
          meta={[
            { charSet: 'utf-8' },
            {
              content: 'Test description',
              name: 'description',
            },
            {
              content: 'text/html',
              httpEquiv: 'content-type',
            },
            { content: 'article', property: 'og:type' },
            {
              content: 'Test name itemprop',
              itemProp: 'name',
            },
          ]}
        />,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];

      expect(existingTags).toBeDefined();

      const filteredTags = existingTags.filter(
        // eslint-disable-next-line jest/no-conditional-in-test
        (tag) => tag.getAttribute('charset') === 'utf-8'
          || (
            // eslint-disable-next-line jest/no-conditional-in-test
            tag.getAttribute('name') === 'description'
            && tag.getAttribute('content') === 'Test description'
          )
          || (
            // eslint-disable-next-line jest/no-conditional-in-test
            tag.getAttribute('http-equiv') === 'content-type'
            && tag.getAttribute('content') === 'text/html'
          )
          || (
            // eslint-disable-next-line jest/no-conditional-in-test
            tag.getAttribute('itemprop') === 'name'
            && tag.getAttribute('content') === 'Test name itemprop'
          ),
      );

      expect(filteredTags.length).toBeGreaterThanOrEqual(4);
    });

    it('clears all meta tags if none are specified', () => {
      renderClient(<Helmet meta={[{ content: 'Test description', name: 'description' }]} />);

      renderClient(<Helmet />);

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it('tags without \'name\', \'http-equiv\', \'property\', \'charset\', or \'itemprop\' are not accepted', () => {
      renderClient(
        <Helmet
          meta={[
            {
              // @ts-expect-error "pre-existing"
              href: 'won\'t work',
            },
          ]}
        />,
      );

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it('sets meta tags based on deepest nested component', () => {
      renderClient(
        <div>
          <Helmet
            meta={[
              { charSet: 'utf-8' },
              {
                content: 'Test description',
                name: 'description',
              },
            ]}
          />
          <Helmet
            meta={[
              {
                content: 'Inner description',
                name: 'description',
              },
              { content: 'test,meta,tags', name: 'keywords' },
            ]}
          />
        </div>,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag, thirdTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(3);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag!.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('charset', 'utf-8');
      expect(firstTag?.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag!.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Inner description');
      expect(secondTag?.outerHTML).toMatchSnapshot();

      expect(thirdTag).toBeInstanceOf(Element);
      expect(thirdTag!.getAttribute).toBeDefined();
      expect(thirdTag).toHaveAttribute('name', 'keywords');
      expect(thirdTag).toHaveAttribute('content', 'test,meta,tags');
      expect(thirdTag?.outerHTML).toMatchSnapshot();
    });

    it('allows duplicate meta tags if specified in the same component', () => {
      renderClient(
        <Helmet
          meta={[
            {
              content: 'Test description',
              name: 'description',
            },
            {
              content: 'Duplicate description',
              name: 'description',
            },
          ]}
        />,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag!.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Test description');
      expect(firstTag?.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag!.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Duplicate description');
      expect(secondTag?.outerHTML).toMatchSnapshot();
    });

    it('overrides duplicate meta tags with single meta tag in a nested component', () => {
      renderClient(
        <div>
          <Helmet
            meta={[
              {
                content: 'Test description',
                name: 'description',
              },
              {
                content: 'Duplicate description',
                name: 'description',
              },
            ]}
          />
          <Helmet
            meta={[
              {
                content: 'Inner description',
                name: 'description',
              },
            ]}
          />
        </div>,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag!.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Inner description');
      expect(firstTag?.outerHTML).toMatchSnapshot();
    });

    it('overrides single meta tag with duplicate meta tags in a nested component', () => {
      renderClient(
        <div>
          <Helmet
            meta={[
              {
                content: 'Test description',
                name: 'description',
              },
            ]}
          />
          <Helmet
            meta={[
              {
                content: 'Inner description',
                name: 'description',
              },
              {
                content: 'Inner duplicate description',
                name: 'description',
              },
            ]}
          />
        </div>,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag!.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Inner description');
      expect(firstTag?.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag!.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Inner duplicate description');
      expect(secondTag?.outerHTML).toMatchSnapshot();
    });

    it('does not render tag when primary attribute is null', () => {
      renderClient(
        <Helmet
          meta={[
            {
              content: 'Inner duplicate description',
              name: undefined,
            },
          ]}
        />,
      );

      const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toHaveLength(0);
    });

    it('fails gracefully when meta is wrong shape', () => {
      const originalConsole = global.console;
      // TODO: Revise.
      // eslint-disable-next-line jest/prefer-spy-on
      global.console.warn = jest.fn();

      renderClient(
        <Helmet
          meta={
            // @ts-expect-error "pre-existing"
            { content: 'some title', name: 'title' }
          }
        />,
      );

      const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toHaveLength(0);

      // TODO: Revise.
      // eslint-disable-next-line jest/prefer-called-with
      expect(console.warn).toHaveBeenCalled();

      // TODO: Revise
      // eslint-disable-next-line jest/prefer-jest-mocked
      expect((console.warn as jest.Mock<unknown, unknown[]>).mock.calls[0]?.[0])
        .toMatchSnapshot();

      global.console = originalConsole;
    });
  });

  describe('Declarative API', () => {
    it('updates meta tags', () => {
      renderClient(
        <Helmet>
          <meta charSet="utf-8" />
          <meta content="Test description" name="description" />
          <meta content="text/html" httpEquiv="content-type" />
          <meta content="article" property="og:type" />
          <meta content="Test name itemprop" itemProp="name" />
        </Helmet>,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];

      expect(existingTags).toBeDefined();

      const filteredTags = existingTags.filter(
        // eslint-disable-next-line jest/no-conditional-in-test
        (tag) => tag.getAttribute('charset') === 'utf-8'
          || (
            // eslint-disable-next-line jest/no-conditional-in-test
            tag.getAttribute('name') === 'description'
            && tag.getAttribute('content') === 'Test description'
          )
          || (
            // eslint-disable-next-line jest/no-conditional-in-test
            tag.getAttribute('http-equiv') === 'content-type'
            && tag.getAttribute('content') === 'text/html'
          )
          || (
            // eslint-disable-next-line jest/no-conditional-in-test
            tag.getAttribute('itemprop') === 'name'
            && tag.getAttribute('content') === 'Test name itemprop'
          ),
      );

      expect(filteredTags.length).toBeGreaterThanOrEqual(4);
    });

    it('clears all meta tags if none are specified', () => {
      renderClient(
        <Helmet>
          <meta content="Test description" name="description" />
        </Helmet>,
      );

      renderClient(<Helmet />);

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it('tags without \'name\', \'http-equiv\', \'property\', \'charset\', or \'itemprop\' are not accepted', () => {
      renderClient(
        <Helmet>
          <meta
            // @ts-expect-error "pre-existing"
            href="won't work"
          />
        </Helmet>,
      );

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it('sets meta tags based on deepest nested component', () => {
      renderClient(
        <div>
          <Helmet>
            <meta charSet="utf-8" />
            <meta content="Test description" name="description" />
          </Helmet>
          <Helmet>
            <meta content="Inner description" name="description" />
            <meta content="test,meta,tags" name="keywords" />
          </Helmet>
        </div>,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag, thirdTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(3);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag!.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('charset', 'utf-8');
      expect(firstTag?.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag!.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Inner description');
      expect(secondTag?.outerHTML).toMatchSnapshot();

      expect(thirdTag).toBeInstanceOf(Element);
      expect(thirdTag!.getAttribute).toBeDefined();
      expect(thirdTag).toHaveAttribute('name', 'keywords');
      expect(thirdTag).toHaveAttribute('content', 'test,meta,tags');
      expect(thirdTag?.outerHTML).toMatchSnapshot();
    });

    it('allows duplicate meta tags if specified in the same component', () => {
      renderClient(
        <Helmet>
          <meta content="Test description" name="description" />
          <meta content="Duplicate description" name="description" />
        </Helmet>,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag!.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Test description');
      expect(firstTag?.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag!.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Duplicate description');
      expect(secondTag?.outerHTML).toMatchSnapshot();
    });

    it('overrides duplicate meta tags with single meta tag in a nested component', () => {
      renderClient(
        <div>
          <Helmet>
            <meta content="Test description" name="description" />
            <meta content="Duplicate description" name="description" />
          </Helmet>
          <Helmet>
            <meta content="Inner description" name="description" />
          </Helmet>
        </div>,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag!.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Inner description');
      expect(firstTag?.outerHTML).toMatchSnapshot();
    });

    it('overrides single meta tag with duplicate meta tags in a nested component', () => {
      renderClient(
        <div>
          <Helmet>
            <meta content="Test description" name="description" />
          </Helmet>
          <Helmet>
            <meta content="Inner description" name="description" />
            <meta content="Inner duplicate description" name="description" />
          </Helmet>
        </div>,
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag!.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Inner description');
      expect(firstTag?.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag!.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Inner duplicate description');
      expect(secondTag?.outerHTML).toMatchSnapshot();
    });

    it('does not render tag when primary attribute is null', () => {
      renderClient(
        <Helmet>
          <meta content="Inner duplicate description" name={undefined} />
        </Helmet>,
      );

      const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toHaveLength(0);
    });
  });
});
