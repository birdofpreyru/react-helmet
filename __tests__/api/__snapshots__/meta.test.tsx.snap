// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`meta tags API allows duplicate meta tags if specified in the same component 1`] = `"<meta content="Test description" name="description" data-rh="true">"`;

exports[`meta tags API allows duplicate meta tags if specified in the same component 2`] = `"<meta content="Duplicate description" name="description" data-rh="true">"`;

exports[`meta tags API fails gracefully when meta is wrong shape 1`] = `"Helmet: meta should be of type "Array". Instead found type "object""`;

exports[`meta tags API overrides duplicate meta tags with single meta tag in a nested component 1`] = `"<meta content="Inner description" name="description" data-rh="true">"`;

exports[`meta tags API overrides single meta tag with duplicate meta tags in a nested component 1`] = `"<meta content="Inner description" name="description" data-rh="true">"`;

exports[`meta tags API overrides single meta tag with duplicate meta tags in a nested component 2`] = `"<meta content="Inner duplicate description" name="description" data-rh="true">"`;

exports[`meta tags API sets meta tags based on deepest nested component 1`] = `"<meta charset="utf-8" data-rh="true">"`;

exports[`meta tags API sets meta tags based on deepest nested component 2`] = `"<meta content="Inner description" name="description" data-rh="true">"`;

exports[`meta tags API sets meta tags based on deepest nested component 3`] = `"<meta content="test,meta,tags" name="keywords" data-rh="true">"`;

exports[`meta tags Declarative API allows duplicate meta tags if specified in the same component 1`] = `"<meta content="Test description" name="description" data-rh="true">"`;

exports[`meta tags Declarative API allows duplicate meta tags if specified in the same component 2`] = `"<meta content="Duplicate description" name="description" data-rh="true">"`;

exports[`meta tags Declarative API overrides duplicate meta tags with single meta tag in a nested component 1`] = `"<meta content="Inner description" name="description" data-rh="true">"`;

exports[`meta tags Declarative API overrides single meta tag with duplicate meta tags in a nested component 1`] = `"<meta content="Inner description" name="description" data-rh="true">"`;

exports[`meta tags Declarative API overrides single meta tag with duplicate meta tags in a nested component 2`] = `"<meta content="Inner duplicate description" name="description" data-rh="true">"`;

exports[`meta tags Declarative API sets meta tags based on deepest nested component 1`] = `"<meta charset="utf-8" data-rh="true">"`;

exports[`meta tags Declarative API sets meta tags based on deepest nested component 2`] = `"<meta content="Inner description" name="description" data-rh="true">"`;

exports[`meta tags Declarative API sets meta tags based on deepest nested component 3`] = `"<meta content="test,meta,tags" name="keywords" data-rh="true">"`;
